cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,316 @@
+/*
+Copyright 2020 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package do
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/digitalocean/godo"
+	"github.com/google/go-cmp/cmp"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+const (
+	// Interval of synchronizing service status from apiserver.
+	serviceSyncPeriod = 30 * time.Second
+	// Interval for frequency of running the firewall controller.
+	controllerLoopRunPeriod = 90 * time.Second
+	// Frequency at which the firewall controller runs.
+	firewallReconcileRequency = 5 * time.Minute
+)
+
+// firewallCache stores a cached firewall and mutex to handle concurrent access.
+type firewallCache struct {
+	mu       *sync.RWMutex // protects firewall.
```

Comment:
With this new change, I get a `copy lock` warning when I remove the mutex:
```
call of newFakeFirewallManagerOp copies lock value: github.com/digitalocean/digitalocean-cloud-controller-manager/cloud-controller-manager/do.firewallCache contains sync.RWMutexcopylocks
```

this line in the `TestFirewallController_Get` tests: 
```go
fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
```

This is due to how my test function is written:
```go
func newFakeFirewallManagerOp(client *godo.Client, cache firewallCache) *firewallManagerOp {
	return &firewallManagerOp{
		client:             client,
		fwCache:            cache,
		workerFirewallName: fakeWorkerFWName,
	}
}
```

Reply:
@timoreimann ^

Reply:
Fixed it 👍. Passing cache in as a pointer to firewallCache, seems to do the trick!

Reply:
One alternative I can think of is to make `firewallManagerOp.firewallCache` a pointer, but there are more changes that need to be made with that option than there is just leaving `mu *sync.RWMutex`. I'm going to leave it with the pointer to `*sync.RWMutex` because it seems like less code changes and pointers being passed around. Let me know what you think.

Reply:
It looks like we could change `newFakeFirewallManagerOp` to take a firewall (as opposed to a firewall cache) and let the implementation populate the cache with the passed firewall. That way, we shouldn't need to pass the entire `firewallCache` variable along and thus prevent the need to copy the mutex.

I'm also okay with keeping things the way are right now.

Reply:
Yeah, let's keep things how they are now and revisit this if we need to in future PR(s).